controllers/utils/failuredomains.go

```diff
@@ -0,0 +1,139 @@
+		Spec:       fdSpec,
+	}
+	return errors.Wrap(r.K8sClient.Create(r.RequestCtx, csFD), "creating CloudStackFailureDomain")
+}
+
+// CreateFailureDomains creates a CloudStackFailureDomain CRD for each of the ReconcilationSubject's FailureDomains.
+func (r *ReconciliationRunner) CreateFailureDomains(fdSpecs []infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		for _, fdSpec := range fdSpecs {
+			if !strings.HasSuffix(fdSpec.Name, "-"+r.CAPICluster.Name) { // Add cluster name suffix if missing.
+				fdSpec.Name = fdSpec.Name + "-" + r.CAPICluster.Name
+			}
+			if err := r.CreateFailureDomain(fdSpec); err != nil {
+				if !strings.Contains(strings.ToLower(err.Error()), "already exists") {
+					return reconcile.Result{}, errors.Wrap(err, "creating CloudStackFailureDomains")
+				}
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// GetFailureDomains gets CloudStackFailureDomains owned by a CloudStackCluster.
+func (r *ReconciliationRunner) GetFailureDomains(fds *infrav1.CloudStackFailureDomainList) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		capiClusterLabel := map[string]string{clusterv1.ClusterLabelName: r.CSCluster.GetLabels()[clusterv1.ClusterLabelName]}
+		if err := r.K8sClient.List(
+			r.RequestCtx,
+			fds,
+			client.InNamespace(r.Request.Namespace),
+			client.MatchingLabels(capiClusterLabel),
+		); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to list failure domains")
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// GetFailureDomainsAndRequeueIfMissing gets CloudStackFailureDomains owned by a CloudStackCluster and requeues if none are found.
+func (r *ReconciliationRunner) GetFailureDomainsAndRequeueIfMissing(fds *infrav1.CloudStackFailureDomainList) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		if res, err := r.GetFailureDomains(fds)(); r.ShouldReturn(res, err) {
+			return res, err
+		} else if len(fds.Items) < 1 {
+			return r.RequeueWithMessage("no failure domains found, requeueing")
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+type CloudClientExtension interface {
+	RegisterExtension(*ReconciliationRunner) CloudClientExtension
+	AsFailureDomainUser(*infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod
+}
+
+type CloudClientImplementation struct {
+	CloudClientExtension
+	*ReconciliationRunner
+}
+
+func (c *CloudClientImplementation) RegisterExtension(r *ReconciliationRunner) CloudClientExtension {
+	c.ReconciliationRunner = r
+	return c
+}
+
+// AsFailureDomainUser uses the credentials specified in the failure domain to set the ReconciliationSubject's CSUser client.
+func (c *CloudClientImplementation) AsFailureDomainUser(fdSpec *infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		if !strings.HasSuffix(fdSpec.Name, "-"+c.CAPICluster.Name) { // Add cluster name suffix if missing.
+			fdSpec.Name = fdSpec.Name + "-" + c.CAPICluster.Name
```

Comment:
formatting is unpleasant here too

Reply:
And here this is still what go fmt wants.